File: crdt/hlc/hlc.go
// Package hlc implements a Hybrid Logical Clock (HLC) for distributed
// causality tracking.
//
// An [HLC] timestamp combines a physical wall-clock component with a logical
// counter and a node identifier, providing a total ordering of events across
// nodes that is consistent with real time. When two events share the same wall
// time, the logical counter breaks ties; when both are equal, the node ID
// provides a deterministic tiebreaker.
//
// Use [NewClock] to create a per-node clock, [Clock.Now] to generate a new
// timestamp, and [Clock.Update] to advance the clock when receiving a remote
// timestamp (ensuring the local clock is always at least as recent as any
// observed remote event).
package hlc

import (
	"fmt"
	"math"
	"sync"
	"time"
)

// HLC represents a Hybrid Logical Clock timestamp.
type HLC struct {
	WallTime int64  `json:"w"` // Physical time (unix nanoseconds)
	Logical  int32  `json:"l"` // Logical counter
	NodeID   string `json:"n"` // Unique node identifier
}

// Compare returns -1 if h < other, 1 if h > other, and 0 if equal.
func (h HLC) Compare(other HLC) int {
	if h.WallTime < other.WallTime {
		return -1
	}
	if h.WallTime > other.WallTime {
		return 1
	}
	if h.Logical < other.Logical {
		return -1
	}
	if h.Logical > other.Logical {
		return 1
	}
	if h.NodeID < other.NodeID {
		return -1
	}
	if h.NodeID > other.NodeID {
		return 1
	}
	return 0
}

// After returns true if h is strictly after other.
func (h HLC) After(other HLC) bool {
	return h.Compare(other) > 0
}

func (h HLC) String() string {
	return fmt.Sprintf("%d:%d:%s", h.WallTime, h.Logical, h.NodeID)
}

// Clock manages the local HLC state.
//
// Latest is exposed for serialisation but must not be mutated by callers
// after the clock is in use — direct writes bypass the internal mutex and
// race with concurrent Now/Update. Use [Clock.SetLatest] for explicit
// rehydration (e.g. from snapshots).
type Clock struct {
	mu     sync.Mutex
	Latest HLC
	NodeID string

	// seq allocates identifiers for sequence elements. It is deliberately
	// independent of Latest: see ReserveSequence.
	seq HLC
}

// NewClock creates a new clock for the given node ID.
func NewClock(nodeID string) *Clock {
	return &Clock{
		NodeID: nodeID,
		Latest: HLC{
			WallTime: 0,
			Logical:  0,
			NodeID:   nodeID,
		},
	}
}

// SetLatest rehydrates the clock from a previously observed timestamp under
// the clock's mutex, so it is safe to call alongside concurrent Now/Update.
// Subsequent Now/Update calls advance from at least h.
func (c *Clock) SetLatest(h HLC) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.Latest = h
}

// Now returns the current HLC timestamp.
func (c *Clock) Now() HLC {
	return c.Reserve(1)
}

// Reserve returns the current HLC timestamp and reserves n logical ticks.
//
// n must be non-negative and small enough that c.Latest.Logical + n fits in
// int32; otherwise Reserve panics. Practical text inserts and similar uses
// fall well under that bound; overflow would silently break causal ordering,
// so an explicit panic is preferred to a wraparound bug.
func (c *Clock) Reserve(n int) HLC {
	if n < 0 {
		panic("hlc: Reserve called with negative n")
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	physNow := time.Now().UnixNano()

	if physNow > c.Latest.WallTime {
		c.Latest.WallTime = physNow
		c.Latest.Logical = 0
	}

	if int64(c.Latest.Logical)+int64(n) > int64(math.MaxInt32) {
		panic("hlc: Reserve would overflow Logical (int32)")
	}

	start := c.Latest
	c.Latest.Logical += int32(n)
	return start
}

// ReserveSequence reserves n consecutive identifiers for the elements of a
// sequence — the characters of a Text, the entries of a List.
//
// Unlike [Clock.Reserve] it does not follow the wall clock. Reserve begins a
// new logical range whenever physical time has moved on, which it always has by
// the next keystroke, so no two identifiers it hands out are ever adjacent and
// a sequence cannot tell that consecutive elements belong together — it stores
// one run per call. ReserveSequence counts on from where it left off, so
// consecutive calls return adjacent blocks and a sequence can hold them as a
// single run. That is the difference between a document costing a hundred bytes
// per character and one costing a handful, and between an edit that walks every
// character and one that walks a few runs.
//
// A sequence asks only that its identifiers be unique and totally ordered, and
// these are both: an element's position comes from the element it follows, and
// the identifier serves only to break ties between insertions made concurrently
// at the same place. These identifiers carry no causal information — use
// [Clock.Now] for a timestamp.
func (c *Clock) ReserveSequence(n int) HLC {
	if n < 0 {
		panic("hlc: ReserveSequence called with negative n")
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	if c.seq.WallTime == 0 {
		// Anchor the range at the current time on first use, so that
		// identifiers issued before a restart cannot be issued again after one.
		c.seq = HLC{WallTime: time.Now().UnixNano(), NodeID: c.NodeID}
	}
	if int64(c.seq.Logical)+int64(n) > int64(math.MaxInt32) {
		panic("hlc: ReserveSequence would overflow Logical (int32)")
	}

	start := c.seq
	c.seq.Logical += int32(n)
	return start
}

// Update updates the local clock based on a remote timestamp.
func (c *Clock) Update(remote HLC) {
	c.mu.Lock()
	defer c.mu.Unlock()

	physNow := time.Now().UnixNano()

	nextWall := physNow
	if c.Latest.WallTime > nextWall {
		nextWall = c.Latest.WallTime
	}
	if remote.WallTime > nextWall {
		nextWall = remote.WallTime
	}

	var nextLog int32
	if nextWall == c.Latest.WallTime && nextWall == remote.WallTime {
		maxLog := c.Latest.Logical
		if remote.Logical > maxLog {
			maxLog = remote.Logical
		}
		nextLog = maxLog + 1
	} else if nextWall == c.Latest.WallTime {
		nextLog = c.Latest.Logical + 1
	} else if nextWall == remote.WallTime {
		nextLog = remote.Logical + 1
	} else {
		nextLog = 0
	}

	c.Latest.WallTime = nextWall
	c.Latest.Logical = nextLog
}

File: js/src/crdt/hlc.ts
/**
 * Hybrid logical clocks, as deep's CRDT uses them.
 *
 * Wall times are Unix **nanoseconds**, which run past 2^53 and so cannot be
 * held in a JavaScript number without losing the low digits — two edits a
 * microsecond apart would land on the same identifier. They are `bigint`
 * here for that reason, all the way through the codec.
 */
export interface HLC {
  /** Wall time, Unix nanoseconds. */
  w: bigint;
  /** Logical counter, breaking ties within a wall time. */
  l: number;
  /** The node that issued it. */
  n: string;
}

export const ZERO: HLC = { w: 0n, l: 0, n: '' };

export function isZero(h: HLC | undefined): boolean {
  return h === undefined || (h.w === 0n && h.l === 0 && h.n === '');
}

/** Orders two identifiers: wall time, then counter, then node id. */
export function compare(a: HLC, b: HLC): number {
  if (a.w !== b.w) return a.w < b.w ? -1 : 1;
  if (a.l !== b.l) return a.l < b.l ? -1 : 1;
  if (a.n !== b.n) return a.n < b.n ? -1 : 1;
  return 0;
}

export function after(a: HLC, b: HLC): boolean {
  return compare(a, b) > 0;
}

export function equal(a: HLC, b: HLC): boolean {
  return a.w === b.w && a.l === b.l && a.n === b.n;
}

/** A map key for an identifier, since objects cannot be keys by value. */
export function key(h: HLC): string {
  return `${h.n} ${h.w} ${h.l}`;
}

/** The identifier `offset` characters after h. */
export function plus(h: HLC, offset: number): HLC {
  return { w: h.w, l: h.l + offset, n: h.n };
}

/**
 * The space an identifier was allocated from: a node together with the point
 * it started counting. A node that restarts allocates from a fresh space, so
 * identifiers issued before a restart can never be issued again after one.
 */
export function origin(h: HLC): string {
  return `${h.n}@${h.w}`;
}

/**
 * Clock allocates identifiers for one node.
 *
 * Sequence identifiers come from their own range, anchored once at the
 * clock's wall time — the counter only ever climbs within that range, which
 * is what lets a state vector describe "everything this node wrote" as a
 * single number.
 */
export class Clock {
  readonly nodeID: string;
  private seq: HLC;

  constructor(nodeID: string, wallTime?: bigint) {
    this.nodeID = nodeID;
    const w = wallTime ?? BigInt(Date.now()) * 1_000_000n;
    this.seq = { w, l: 0, n: nodeID };
  }

  /** Reserves n consecutive identifiers and returns the first. */
  reserveSequence(n: number): HLC {
    const start = { ...this.seq };
    this.seq = { ...this.seq, l: this.seq.l + n };
    return start;
  }
}
